domain/consensus/processes/blockparentbuilder/blockparentbuilder.go

```diff
@@ -0,0 +1,214 @@
+		blockHeaderStore:      blockHeaderStore,
+		dagTopologyManager:    dagTopologyManager,
+		reachabilityDataStore: reachabilityDataStore,
+		pruningStore:          pruningStore,
+	}
+}
+
+func (bpb *blockParentBuilder) BuildParents(stagingArea *model.StagingArea,
+	directParentHashes []*externalapi.DomainHash) ([]externalapi.BlockLevelParents, error) {
+
+	pruningPoint, err := bpb.pruningStore.PruningPoint(bpb.databaseContext, stagingArea)
+	if err != nil {
+		return nil, err
+	}
+
+	directParentHeaders := make(map[externalapi.DomainHash]externalapi.BlockHeader)
+	for _, directParentHash := range directParentHashes {
+		directParentHeader, err := bpb.blockHeaderStore.BlockHeader(bpb.databaseContext, stagingArea, directParentHash)
+		if err != nil {
+			return nil, err
+		}
+		directParentHeaders[*directParentHash] = directParentHeader
+	}
+
+	// Direct parents are guaranteed to be in one other's anticones so add them all to
+	// all the block levels they occupy
+	candidatesByLevelToReferenceBlocksMap := make(map[int]map[externalapi.DomainHash][]*externalapi.DomainHash)
+
+	for directParentHash, directParentHeader := range directParentHeaders {
+		directParentHash := directParentHash // Assign to a new pointer to avoid `range` pointer reuse
+		proofOfWorkValue := pow.CalculateProofOfWorkValue(directParentHeader.ToMutable())
+		for blockLevel := 0; ; blockLevel++ {
+			if _, exists := candidatesByLevelToReferenceBlocksMap[blockLevel]; !exists {
+				candidatesByLevelToReferenceBlocksMap[blockLevel] = make(map[externalapi.DomainHash][]*externalapi.DomainHash)
+			}
+			candidatesByLevelToReferenceBlocksMap[blockLevel][directParentHash] = []*externalapi.DomainHash{&directParentHash}
+			if proofOfWorkValue.Bit(blockLevel+1) != 0 {
+				break
+			}
+		}
+	}
+
+	virtualGenesisChildren, err := bpb.dagTopologyManager.Children(stagingArea, model.VirtualGenesisBlockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	virtualGenesisChildrenHeaders := make(map[externalapi.DomainHash]externalapi.BlockHeader, len(virtualGenesisChildren))
+	for _, child := range virtualGenesisChildren {
+		virtualGenesisChildrenHeaders[*child], err = bpb.blockHeaderStore.BlockHeader(bpb.databaseContext, stagingArea, child)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	maybeAddDirectParentParents := func(directParentHeader externalapi.BlockHeader) error {
+		for blockLevel, blockLevelParentsInHeader := range directParentHeader.Parents() {
+			isEmptyLevel := false
+			if _, exists := candidatesByLevelToReferenceBlocksMap[blockLevel]; !exists {
+				candidatesByLevelToReferenceBlocksMap[blockLevel] = make(map[externalapi.DomainHash][]*externalapi.DomainHash)
+				isEmptyLevel = true
+			}
+
+			for _, parent := range blockLevelParentsInHeader {
+				hasReachabilityData, err := bpb.reachabilityDataStore.HasReachabilityData(bpb.databaseContext, stagingArea, parent)
+				if err != nil {
+					return err
+				}
+
+				var referenceBlocks []*externalapi.DomainHash
```

Comment:
Please add a comment explaining why we do A when we have reachability data and why we do B when we don't